oonib/testhelpers/http_helpers.py

```diff
@@ -168,7 +241,336 @@ def all(self, length, keyword):
+        lifecycle = self._cache_lifecycle[response_type][key_hash]
+        if lifecycle['expiration'].active():
+            lifecycle['expiration'].cancel()
+
+        yield lifecycle['lock'].acquire()
+        try:
+            os.remove(os.path.join(self._cache_dir, response_type, key_hash))
+        except OSError:
+            pass
+        lifecycle['lock'].release()
+        del self._cache_lifecycle[response_type][key_hash]
+
+    @defer.inlineCallbacks
+    def lookup(self, response_type, key):
+        if not self.enable_caching:
+            defer.returnValue(None)
+
+        key_hash = sha256(key).hexdigest()
+        cache_file = os.path.join(self._cache_dir, response_type, key_hash)
+
+        if key_hash not in self._cache_lifecycle[response_type]:
+            defer.returnValue(None)
+
+        lock = self._cache_lifecycle[response_type][key_hash]['lock']
+        expiration = \
+            self._cache_lifecycle[response_type][key_hash]['expiration']
+
+        yield lock.acquire()
+
+        if not os.path.exists(cache_file):
+            lock.release()
+            defer.returnValue(None)
+
+        with open(cache_file, 'r') as fh:
+            value = json.load(fh)
+
+        expiration.reset(self.expiration_time)
+        lock.release()
+        defer.returnValue(value)
+
+    @defer.inlineCallbacks
+    def http_request(self, url):
+        cached_value = yield self.lookup('http_request', url)
+        if cached_value is not None:
+            defer.returnValue(cached_value)
+
+        page_info = {
+            'body_length': -1,
+            'status_code': -1,
+            'headers': {},
+            'failure': None
+        }
+
+        agent = ContentDecoderAgent(FixedRedirectAgent(Agent(reactor)),
+                                    [('gzip', GzipDecoder)])
+        try:
+            retries = 0
+            while True:
+                try:
+                    response = yield agent.request('GET', url,
+                                                   TrueHeaders(REQUEST_HEADERS))
+                    headers = {}
+                    for name, value in response.headers.getAllRawHeaders():
+                        headers[name] = unicode(value[0], errors='ignore')
+                    body_length = -1
+                    try:
+                        body = yield readBody(response)
+                        body_length = len(body)
+                    except PartialDownloadError as pde:
+                        if pde.response:
+                            body_length = len(pde.response)
+                    page_info['body_length'] = body_length
+                    page_info['status_code'] = response.code
+                    page_info['headers'] = headers
+                    page_info['title'] = extractTitle(body)
+                    if self.enable_debug:
+                        response.body = body
+                        page_info['responses'] = encodeResponses(response)
```

Comment:
I think I missed so far whether this option of enabling debug could be enabled by client action or is a server side only property. I see cases whether it would be useful to ask the backend to retrieve more. Thoughts?